src/Core/Grid/Query/StoreQueryBuilder.php

```diff
@@ -0,0 +1,177 @@
+<?php
+/**
+ * Copyright since 2007 PrestaShop SA and Contributors
+ * PrestaShop is an International Registered Trademark & Property of PrestaShop SA
+ *
+ * NOTICE OF LICENSE
+ *
+ * This source file is subject to the Open Software License (OSL 3.0)
+ * that is bundled with this package in the file LICENSE.md.
+ * It is also available through the world-wide-web at this URL:
+ * https://opensource.org/licenses/OSL-3.0
+ * If you did not receive a copy of the license and are unable to
+ * obtain it through the world-wide-web, please send an email
+ * to license@prestashop.com so we can send you a copy immediately.
+ *
+ * DISCLAIMER
+ *
+ * Do not edit or add to this file if you wish to upgrade PrestaShop to newer
+ * versions in the future. If you wish to customize PrestaShop for your
+ * needs please refer to https://devdocs.prestashop.com/ for more information.
+ *
+ * @author    PrestaShop SA and Contributors <contact@prestashop.com>
+ * @copyright Since 2007 PrestaShop SA and Contributors
+ * @license   https://opensource.org/licenses/OSL-3.0 Open Software License (OSL 3.0)
+ */
+declare(strict_types=1);
+
+namespace PrestaShop\PrestaShop\Core\Grid\Query;
+
+use Doctrine\DBAL\Connection;
+use Doctrine\DBAL\Query\QueryBuilder;
+use PrestaShop\PrestaShop\Core\Grid\Search\SearchCriteriaInterface;
+
+class StoreQueryBuilder extends AbstractDoctrineQueryBuilder
+{
+    /**
+     * @var DoctrineSearchCriteriaApplicatorInterface
+     */
+    private $searchCriteriaApplicator;
+
+    /**
+     * @var int[]
+     */
+    protected $shopIds;
+
+    /**
+     * @var int
+     */
+    protected $languageId;
+
+    /**
+     * @param Connection $connection
+     * @param string $dbPrefix
+     * @param DoctrineSearchCriteriaApplicatorInterface $searchCriteriaApplicator
+     * @param int[] $shopIds
+     * @param int $languageId
+     */
+    public function __construct(
+        Connection $connection,
+        string $dbPrefix,
+        DoctrineSearchCriteriaApplicatorInterface $searchCriteriaApplicator,
+        array $shopIds,
```

Comment:
Instead of injecting the shop Ids directly in this query builder from the shop context you could change the `StoreFilters` so that it extends `PrestaShop\PrestaShop\Core\Search\ShopFilters` (similar to what was done in product grid). The parameter resolver then automatically inject a `ShopConstraint` based on the context for now (but will be improved in the future to handle GET parameters).

This way you get a `ShopSearchCriteriaInterface` as a parameter that gives you all the details to correctly build your query. If possible future grid should use this as a base so that piece by piece we can remove the direct dependencies to ShopContext.

Although there is room for improvement, starting with the interface that so far only handles `SearchCriteriaInterface` I wonder if we could extend `DoctrineQueryBuilderInterface` in another interface that would accept `ShopSearchCriteriaInterface` as a parameter 🤔

Reply:
I added it, though it seems such an onverhead to extract shopIds from constraint :exploding_head: